src/wallet.ts

```diff
@@ -1,31 +1,206 @@
+}
 
-export const generateEmptyWallet = () => {
+const generateEmptyWallet = () => {
     const privateKey = secp256k1.generatePrivateKey()
     const publicKey = secp256k1.derivePublicKey(privateKey)
-    const addr = address.fromPublicKey(publicKey)
+    const addr = addressUtils.fromPublicKey(publicKey).toLowerCase()
 
     return {
         privateKey: privateKey.toString('hex'),
         address: addr,
     }
 }
 
-export const generateWalletWithFunds = async () => {
-    const wallet = generateEmptyWallet()
-    const receipt = await fundAccount(wallet.address)
+type WaitForFunding = () => Promise<
+    components['schemas']['GetTxReceiptResponse'] | undefined
+>
 
-    return {
-        ...wallet,
-        ...receipt,
+class ThorWallet {
+    public readonly address: string
+    public readonly privateKey: Buffer
+    public readonly waitForFunding: WaitForFunding
+
+    constructor(privateKey: Buffer, waitForFunding?: WaitForFunding) {
+        this.privateKey = privateKey
+        this.address = addressFromPrivateKey(privateKey)
+        if (waitForFunding) {
+            this.waitForFunding = waitForFunding
+        } else {
+            this.waitForFunding = () => Promise.resolve(undefined)
+        }
+    }
+
+    public static new(requireFunds: boolean) {
+        const privateKey = secp256k1.generatePrivateKey()
+
+        if (!requireFunds) {
+            return new ThorWallet(privateKey)
+        }
+
+        const addr = addressFromPrivateKey(privateKey)
+
+        const receipt = fundAccount(addr).then((res) => res.receipt)
+
+        return new ThorWallet(privateKey, () => receipt)
+    }
+
+    public deployContract = async (
+        bytecode: string,
+        delegate = false,
+    ): Promise<{
+        contractAddress: string
+        receipt: components['schemas']['GetTxReceiptResponse']
+    }> => {
+        const receipt = await this.sendClauses(
+            [
+                {
+                    to: null,
+                    value: 0,
+                    data: bytecode,
+                },
+            ],
+            true,
+            delegate,
+        )
+
+        const contractAddress = receipt.outputs?.[0].contractAddress
+
+        if (!contractAddress) {
+            throw new Error('Could not get contract address from receipt')
+        }
+
+        return { contractAddress, receipt }
+    }
+
+    public buildTransaction = async (
+        clauses: TransactionClause[],
+    ): Promise<TransactionBody> => {
+        const bestBlockRef = await getBlockRef('best')
+        const genesisBlock = await Node1Client.getBlock('0')
+
+        if (!genesisBlock.success || !genesisBlock.body?.id) {
+            throw new Error('Could not get best block')
+        }
+
+        return {
+            blockRef: bestBlockRef,
+            expiration: 1000,
+            clauses: clauses,
+            gasPriceCoef: 0,
+            gas: 1_000_000,
+            dependsOn: null,
```

Comment:
wonder if we should expose more fields, or allow user to pass own tx?

Reply:
It's not necessary IMO, we can do something like this if we want specific fields:

```typescript
        const clauses = [
            {
                to: wallet.address,
                value: '0x100',
                data: '0x',
            },
        ]

        const txBody = await wallet.buildTransaction(clauses)

        txBody.expiration = -1
        
        const tx = await wallet.signTransaction(new Transaction(txBody))
```